File: examples/javascript/multi_tenant_example.mjs
/** MultiTenantIndex -- isolated vector indices per tenant. */

import { randomUUID } from "node:crypto";
import { MultiTenantIndex } from "membrain";

const DIMENSION = 128;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const index = new MultiTenantIndex({
  dimension: DIMENSION,
  indexType: "hnsw",
  maxTenants: 100,
  indexConfig: { m: 16, ef_construction: 100 },
});

try {
  const tenants = ["team-alpha", "team-beta", "team-gamma"];
  for (const tenant of tenants) index.createTenant(tenant);

  for (const tenant of tenants) {
    const offset = tenants.indexOf(tenant) * 100;
    for (let i = 0; i < 50; i++) index.add(tenant, randomUUID(), embed(offset + i));
  }

  const query = embed(0);
  console.log("Search per tenant (same query):");
  for (const tenant of tenants) {
    const results = index.search(tenant, query, 3);
    const scores = results.map((r) => r.score.toFixed(4));
    console.log(`  ${tenant}: ${index.tenantLen(tenant)} vectors, scores=[${scores}]`);
  }

  index.deleteTenant("team-gamma");
  console.log(`\nAfter deleting team-gamma: ${JSON.stringify(index.listTenants())}`);
} finally {
  index.close();
}

File: examples/python/multi_tenant_example.py
"""MultiTenantIndex -- isolated vector indices per tenant."""

import random
import uuid

from membrain import MultiTenantIndex

DIMENSION = 128


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    with MultiTenantIndex(
        dimension=DIMENSION, index_type="hnsw", max_tenants=100,
        index_config={"m": 16, "ef_construction": 100},
    ) as index:
        tenants = ["team-alpha", "team-beta", "team-gamma"]
        for tenant in tenants:
            index.create_tenant(tenant)

        for tenant in tenants:
            offset = tenants.index(tenant) * 100
            for i in range(50):
                index.add(tenant, str(uuid.uuid4()), embed(seed=float(offset + i)))

        query = embed(seed=0.0)
        print("Search per tenant (same query):")
        for tenant in tenants:
            results = index.search(tenant, query, k=3)
            scores = [f"{r.score:.4f}" for r in results]
            print(f"  {tenant}: {index.tenant_len(tenant)} vectors, scores={scores}")

        index.delete_tenant("team-gamma")
        print(f"\nAfter deleting team-gamma: {index.list_tenants()}")


if __name__ == "__main__":
    main()
